fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java

```diff
@@ -0,0 +1,148 @@
+    private final Statistics rightStatistics;
+
+    private final ExpressionContext context;
+
+    private final List<PhysicalPropertySet> inputProperties;
+
+    private final List<BinaryPredicateOperator> eqOnPredicates;
+
+    public HashJoinCostModel(ExpressionContext context, List<PhysicalPropertySet> inputProperties,
+                             List<BinaryPredicateOperator> eqOnPredicates) {
+        this.context = context;
+        this.leftStatistics = context.getChildStatistics(0);
+        this.rightStatistics = context.getChildStatistics(1);
+        this.inputProperties = inputProperties;
+        this.eqOnPredicates = eqOnPredicates;
+    }
+
+    public double getCpuCost() {
+        String execMode = deriveJoinExecMode();
+        double buildCost;
+        double probeCost;
+        double leftOutput = leftStatistics.getOutputSize(context.getChildOutputColumns(0));
+        double rightOutput = rightStatistics.getOutputSize(context.getChildOutputColumns(1));
+        int parallelFactor = Math.max(ConnectContext.get().getAliveBackendNumber(),
+                ConnectContext.get().getSessionVariable().getDegreeOfParallelism());
+        switch (execMode) {
+            case BROADCAST:
+                buildCost = rightOutput;
+                probeCost = leftOutput * getAvgProbeCost();
+                break;
+            case SHUFFLE:
+                buildCost = rightOutput / parallelFactor;
+                probeCost = leftOutput * getAvgProbeCost();
+                break;
+            default:
+                buildCost = rightOutput;
+                probeCost = leftOutput;
+        }
+        return buildCost + probeCost;
+    }
+
+    public double getMemCost() {
+        String execMode = deriveJoinExecMode();
+        double rightOutput = rightStatistics.getOutputSize(context.getChildOutputColumns(1));
+        double memCost;
+        int beNum = Math.max(1, ConnectContext.get().getAliveBackendNumber());
+
+        if (BROADCAST.equals(execMode)) {
+            memCost = rightOutput * beNum;
+        } else {
+            memCost = rightOutput;
+        }
+        return memCost;
+    }
+
+    private double getAvgProbeCost() {
+        String execMode = deriveJoinExecMode();
+        double keySize = 0;
+        for (BinaryPredicateOperator predicateOperator : eqOnPredicates) {
+            ColumnRefOperator leftCol = (ColumnRefOperator) predicateOperator.getChild(0);
+            ColumnRefOperator rightCol = (ColumnRefOperator) predicateOperator.getChild(1);
+            if (context.getChildStatistics(1).getColumnStatistics().containsKey(leftCol)) {
+                keySize += context.getChildStatistics(1).getColumnStatistic(leftCol).getAverageRowSize();
+            } else if (context.getChildStatistics(1).getColumnStatistics().containsKey(rightCol)) {
+                keySize += context.getChildStatistics(1).getColumnStatistic(rightCol).getAverageRowSize();
+            }
+        }
+        double degradeRatio;
```

Comment:
what's the degrade meaning here?

Reply:
When right table size grows, the probe efficiency will degrade. Use the degradeRation to distinguish the costs for different join situation.

Reply:
I see. but I think maybe we could have a clear name, such as memoryCachePenaltyRatio ?

Reply:
Sounds good.